file_formats/script.md

```diff
@@ -1,101 +1,116 @@
 
-| ID            | Expected Values   | Use |
-|---------------|-------------------|-----|
-| Object		|					| keyword |
-| EndObj		|					| keyword |
-| Nahrung		| fp				| |
-| Soldat		| var, int, int		| cost of soldiers |
-| Turm			| var, int			| cost of tower |
+### Keyword Properties ###
+
+Four keyword properties are known:
+
+| Keyword | Purpose |
+|---------|---------|
+| `Include`| include the contents of another script file |
+| `Objekt` | begin object definition |
+| `EndObj` | end object definition |
+| `ObjFill` | defines default properties for following objects, or recalls properties from a previously defined object |
+
 
 ### Objects ###
 
-**Object Definitions** are in most cases **Arrays** of a defined type.
-**Objects** can be nested (containing itself a Object), two deep.
-The **Arrays** start with the **Keyword** `Objekt` followed by a `:`, whitespace and the **Type** (a **Variable**) of the **Object**.
+Object definitions can take two forms: arrays of a defined object struct-like type, or individual struct-like object definitions.
+
+Object array definitions follow a syntax like the following:
+```
+Objekt: <type>
 
-The next **Object** in the **Array** (??) starts with the `Nummer:` definition.
+    Nummer:     <nummer value or reference>
+    PropertyA:  0
+    PropertyB:  1
 
-The **Array** ends with the **Keyword** `EndObj` usually followed by an `;`(y??).
+    Nummer:     <nummer value or reference>
+    PropertyA:  0
+    PropertyB:  1
 
-The **Objects** can be filled, usually with defaults, with the `ObjFill:` **Keyword**.
-It accepts an **"Nummer" Range** (start and end, comma seperated) or just an **"Nummer"**.
+EndObj;
+```
 
-### Grammar Rules ###
+Struct-like object definitions use a similar syntax without `Nummer` properties:
 
-TODO
 ```
+Objekt:     HAUS_PRODTYP
+    Kind:       ROHSTWACHS
+    Ware:       NOWARE
+    Interval:   300
+EndObj;
 ```
 
-## Notes ##
+Objects can be nested (containing itself an object). The maximum nesting level observed in the game's files is 2.
+
+### Object Defaults / ObjFill ###
+
+The special `ObjFill` property permits reuse of object properties for object arrays. There are two variants: forward-filling and backward-filling.
+
+* Forward-filling defines a default object which includes a statement such as `ObjFill: 0, MAXHAUS`. The semantics of this are to fill array objects in the given range with the properties of that default object.
+* Backward-filling adds the properties of a previously defined object with a certain `Nummer` (i.e. array index) to the current array object. This requires a statement such as `ObjFill: <Nummer reference>`, for example `ObjFill: HAUSWACHS`.
+
+An object can be both forward-filled and backward-filled at the same time.
 
-Scripts usually have inconsistent indention, but use **Spaces** in most cases.
+### EBNF Grammar ###
 
-Some scripts are erroneous (??). Occasionally an extra `@` is present on **Properties**. **Undefined Variables** are used (??).
```

Comment:
you removed the notes. whats the reasoning behind this?

Reply:
The undefined variables aspect is explained in the variables section. In the KE files, I only saw a single extra `@` on properties (the first @Nummer in haeuser.cod), so I figured it would not be necessary to keep this hint. Things might be different for original/NINA/history, so I'll add it back. :slightly_smiling_face:

Reply:
yea, pleas keep it :)
also, do you know if the new version ubisoft release has any changes? I know that they changed the .bsh files a little.

Reply:
btw, at the time i wrote this document, it had been some time since i wrote my parser (which is still not public lol), so all (??) are just facts i was not 100% sure of, that i still remembered. :)

Reply:
Nah, so far, I've only looked at the KE version.

Reply:
> so all (??) are just facts i was not 100% sure of, that i still remembered

All good, it was helpful to have your notes for understanding the file format.